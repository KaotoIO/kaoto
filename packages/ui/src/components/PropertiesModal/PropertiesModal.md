## PropertiesModal

This folder contains `PropertiesModal` component which defines how the properties detail modal with tabs and tables is rendered. The `PropertiesModal` component decides which tabs should be shown according to `CatalogKind` enum (from the input tile). For a particular `CatalogKind`, it uses a particular transform function from `camel-to-tabs.adapter.ts` util to get an array of `IPropertiesTab` for rendering. It passes all that info to the `PropertiesTabs` component which is responsible for rendering tables. The `PropertiesTabs` component renders all tabs from the input array of `IPropertiesTab` and for each tab, it renders all tables which are stored in `IPropertiesTab.tables`. According to a type of table, `Simple` or `Tree` table is rendered.
  
That transformation functions in `camel-to-tabs.adapter.ts` contain a "definition" of tabs for each `CatalogKind` and according to that, it calls particular functions from `camel-to-table.adapter.ts` util to get all `IPropertiesTable` which are related to that `CatalogKind`.

That functions in `camel-to-table.adapter.ts` contain a "definition" of tables.


### How to update table to add a new column

To add a new column, extend models in `PropertiesTable.models.ts` (`PropertiesHeaders` and `IPropertiesRow`), update `camel-to-table.adapter.ts` if the column is needed for a particular definition, and when the data in the cell contins special formation, add new case into `PropertiesTableCommon.tsx`

__Make sure__ that orders of headers in the `IPropertiesTable.headers` match with the orders of element in a particular row `IPropertiesRow`

e.g.
```
{
    headers: [
      PropertiesHeaders.Name,
      PropertiesHeaders.Type,
      PropertiesHeaders.Required
    ],
    rows: [{
        name: "xyz",
        type: "xyz",
        required: "xyz",
    }],
  }
```

### How to add brand new Catalog type

To add a new type of catalog definition, extend the switch in `PropertiesModal` which will cover that new type case. After that create a tab transformation function into `camel-to-tabs.adapter.ts` where define which tabs will be rendered and which tables will be contained. If you need a new table type, add the table transformation function into `camel-to-table.adapter.ts`

### How to add metadata information into table for existing cell

If you need to pass some metadata information, e.g. for formatting existing cells (e.g. if Required==true => add Required as suffix of text), extend `IPropertiesRowAdditionalInfo` object about new information. That information will be available in `PropertiesTableCommon.tsx` where you can define what should be done according to that data.

### Index

- `PropertiesModal` - modal component
- `PropertiesTabs` - tabs component
- `PropertiesTableCommon` - the functions render headers row or data cells row. They contain cases when cell data needs some special formatting according to row metadata (`IPropertiesRow.rowAdditionalInfo`). The functions are common in both, Simple and Tree tables.
- `PropertiesTableSimple` - simple table which render data by rows
- `PropertiesTableTree` - tree table which render data by rows and childrens
- `camel-to-tabs` - functions which define how many tabs will be rendered for a particular catalog type, how they will look and which tables will be contained
- `camel-to-table` - functions which define what data will be in the table according to properties type
