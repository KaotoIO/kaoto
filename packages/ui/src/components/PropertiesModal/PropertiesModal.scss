.properties-modal {
  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v6-c-table__th {
    text-transform: capitalize;
  }

  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v6-c-table__td[data-label='property'] {
    font-weight: bold;
  }

  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v6-c-table__td[data-label='name'] {
    font-weight: bold;
  }

  span[data-label='required'] {
    font-weight: bold;
  }

  span[data-label='autowired'] {
    font-weight: bold;
  }

  span[data-label='group'] {
    font-weight: normal;
  }

  &__body {
    height: 75vh;

    @media screen and (height <= 720px) {
      height: 100vh;
    }

    @media screen and (height <= 1080px) {
      height: 75vh;
    }

    @media screen and (height <= 1440px) {
      height: 60vh;
    }
  }

  &__title-div {
    display: flex;
  }

  &__title-image {
    height: 50px;
    width: 50px;
  }

  &__title {
    padding: 10px;
  }
}
