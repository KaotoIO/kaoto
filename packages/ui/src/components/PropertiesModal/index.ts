export * from './PropertiesModal';
export * from './PropertiesTabs';
export * from './PropertiesTable';
export * from './EmptyTableState';
export * from './PropertiesModal.models';
export * from './PropertiesModal';

