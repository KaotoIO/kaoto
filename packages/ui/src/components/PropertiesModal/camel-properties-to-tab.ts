import { IPropertiesTableFilter } from '../../camel-utils/camel-to-table.adapter';
import { IPropertiesTab, IPropertiesTable } from './PropertiesModal.models';

interface IPropertiesTransformToTable {
  transformFunction: (properties: any, filter?: IPropertiesTableFilter) => IPropertiesTable;
  propertiesObject: any;
  tableCaption?: string;
  filter?: IPropertiesTableFilter;
}

/**
 * This function transforms properties data (from camel-component, camel-processor, kamelets) into one or more tables which are added into the final IPropertiesTab.
 *
 * @param transformationsData object which contains function which is used for transfromation properties object into IPropertiesTable, optionally filter and table caption
 * @param tabTitle - Title of the Tab
 * @returns Tab data which will be use for rendering one Tab
 */
export const transformPropertiesIntoTab = (
  transformationsData: IPropertiesTransformToTable[],
  tabTitle: string,
): IPropertiesTab | undefined => {
  let tables = [];
  for (let transformationData of transformationsData) {
    let table = transformationData.transformFunction(transformationData.propertiesObject, transformationData.filter);
    if (table.rows.length == 0) continue; // we don't care about empty table
    if (transformationData.tableCaption)
      table.caption = transformationData.tableCaption + ' (' + table.rows.length + ')';
    tables.push(table);
  }
  if (tables.length == 0) return undefined;

  let allRowsInAllTables = 0;
  tables.forEach((table) => {
    allRowsInAllTables += table.rows.length;
  });
  return {
    rootName: tabTitle + ' (' + allRowsInAllTables + ')',
    tables: tables,
  };
};
