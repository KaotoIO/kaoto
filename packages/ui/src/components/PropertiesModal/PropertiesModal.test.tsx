import { screen } from '@testing-library/dom';
import { render } from '@testing-library/react';

import { ITile } from '../Catalog';
import { PropertiesModal } from './PropertiesModal';

describe('Component tile', () => {
  const tile: ITile = {
    type: 'component',
    name: 'atom',
    title: 'Atom',
    description: 'Poll Atom RSS feeds.',
    headerTags: ['Stable'],
    tags: ['document', '4.0.0'],
    rawObject: {
      component: {
        kind: 'component',
        name: 'atom',
        title: 'Atom',
        description: 'Poll Atom RSS feeds.',
      },
      properties: {
        bridgeErrorHandler: {
          kind: 'property',
          displayName: 'Bridge Error Handler',
          required: false,
          type: 'boolean',
          defaultValue: false,
          description: 'Allows for bridging the consumer to the Camel',
        },
      },
    },
  };

  it('renders property modal correctly', () => {
    // modal uses React portals so baseElement needs to be used here
    const { baseElement } = render(<PropertiesModal tile={tile} isModalOpen={true} onClose={jest.fn()} />);
    expect(baseElement).toMatchSnapshot();

    // info
    expect(baseElement.getElementsByClassName('pf-v5-c-modal-box__title-text').item(0)).toHaveTextContent('Atom');
    expect(screen.getByTestId('properties-modal-table-description')).toBeInTheDocument();
    expect(screen.getByTestId('properties-modal-table-description')).toHaveTextContent('Poll Atom RSS feeds.');
    expect(screen.getByTestId('properties-modal-table-caption')).toBeInTheDocument();
    expect(screen.getByTestId('properties-modal-table-caption')).toHaveTextContent('Available properties (1)');
    // headers
    expect(screen.getByTestId('header-name')).toBeInTheDocument();
    expect(screen.getByTestId('header-name')).toHaveTextContent('name');
    expect(screen.getByTestId('header-type')).toBeInTheDocument();
    expect(screen.getByTestId('header-type')).toHaveTextContent('type');
    expect(screen.getByTestId('header-kind')).toBeInTheDocument();
    expect(screen.getByTestId('header-kind')).toHaveTextContent('kind');
    expect(screen.getByTestId('header-required')).toBeInTheDocument();
    expect(screen.getByTestId('header-required')).toHaveTextContent('required');
    expect(screen.getByTestId('header-defaultValue')).toBeInTheDocument();
    expect(screen.getByTestId('header-defaultValue')).toHaveTextContent('defaultValue');
    expect(screen.getByTestId('header-description')).toBeInTheDocument();
    expect(screen.getByTestId('header-description')).toHaveTextContent('description');
    // row
    expect(screen.getByTestId('row-0-cell-name')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-name')).toHaveTextContent('Bridge Error Handler');
    expect(screen.getByTestId('row-0-cell-type')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-type')).toHaveTextContent('boolean');
    expect(screen.getByTestId('row-0-cell-kind')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-kind')).toHaveTextContent('property');
    expect(screen.getByTestId('row-0-cell-required')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-required')).toHaveTextContent('false');
    expect(screen.getByTestId('row-0-cell-defaultValue')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-defaultValue')).toHaveTextContent('false');
    expect(screen.getByTestId('row-0-cell-description')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-description')).toHaveTextContent(
      'Allows for bridging the consumer to the Camel',
    );
  });
});

describe('Model tile with empty properties', () => {
  const tile: ITile = {
    type: 'model',
    name: 'apiKey',
    title: 'Api Key',
    description: 'Rest security basic auth definition',
    headerTags: ['Stable'],
    tags: ['document', '4.0.0'],
    rawObject: {
      model: {
        kind: 'model',
        name: 'apiKey',
        title: 'Api Key',
        description: 'Rest security basic auth definition',
      },
      properties: {},
    },
  };

  it('renders property modal correctly', () => {
    // modal uses React portals so baseElement needs to be used here
    const { baseElement } = render(<PropertiesModal tile={tile} isModalOpen={true} onClose={jest.fn()} />);
    expect(baseElement).toMatchSnapshot();

    // info
    expect(baseElement.getElementsByClassName('pf-v5-c-modal-box__title-text').item(0)).toHaveTextContent('Api Key');
    expect(screen.getByTestId('properties-modal-table-description')).toBeInTheDocument();
    expect(screen.getByTestId('properties-modal-table-description')).toHaveTextContent(
      'Rest security basic auth definition',
    );
    expect(screen.getByTestId('properties-modal-table-caption')).toBeInTheDocument();
    expect(screen.getByTestId('properties-modal-table-caption')).toHaveTextContent('Available properties (0)');
    expect(screen.getByTestId('empty-state')).toBeInTheDocument();
  });
});

describe('Kamelet tile', () => {
  const tile: ITile = {
    type: 'kamelet',
    name: 'aws-ddb-streams-source',
    title: 'AWS DynamoDB Streams Source',
    description: 'Receive events from Amazon DynamoDB Streams.',
    headerTags: ['Stable'],
    tags: ['source', '4.0.0-RC1'],
    rawObject: {
      apiVersion: 'camel.apache.org/v1alpha1',
      kind: 'Kamelet',
      metadata: {
        name: 'aws-ddb-streams-source',
      },
      spec: {
        definition: {
          title: 'AWS DynamoDB Streams Source',
          description: 'Receive events',
          required: ['table'],
          type: 'object',
          properties: {
            table: {
              title: 'Table',
              description: 'The name of the DynamoDB table.',
              type: 'string',
            },
            key: {
              title: 'Key',
              description: 'New key',
              type: 'string',
              default: 'amazonaws.com',
            },
          },
        },
      },
    },
  };

  it('renders property modal correctly', () => {
    // modal uses React portals so baseElement needs to be used here
    const { baseElement } = render(<PropertiesModal tile={tile} isModalOpen={true} onClose={jest.fn()} />);
    expect(baseElement).toMatchSnapshot();

    // info
    expect(baseElement.getElementsByClassName('pf-v5-c-modal-box__title-text').item(0)).toHaveTextContent(
      'AWS DynamoDB Streams Source',
    );
    expect(screen.getByTestId('properties-modal-table-description')).toBeInTheDocument();
    expect(screen.getByTestId('properties-modal-table-description')).toHaveTextContent('Receive events');
    expect(screen.getByTestId('properties-modal-table-caption')).toBeInTheDocument();
    expect(screen.getByTestId('properties-modal-table-caption')).toHaveTextContent('Available properties (2)');
    // headers
    expect(screen.getByTestId('header-name')).toBeInTheDocument();
    expect(screen.getByTestId('header-name')).toHaveTextContent('name');
    expect(screen.getByTestId('header-type')).toBeInTheDocument();
    expect(screen.getByTestId('header-type')).toHaveTextContent('type');
    expect(screen.getByTestId('header-required')).toBeInTheDocument();
    expect(screen.getByTestId('header-required')).toHaveTextContent('required');
    expect(screen.getByTestId('header-defaultValue')).toBeInTheDocument();
    expect(screen.getByTestId('header-defaultValue')).toHaveTextContent('defaultValue');
    expect(screen.getByTestId('header-description')).toBeInTheDocument();
    expect(screen.getByTestId('header-description')).toHaveTextContent('description');
    // row
    expect(screen.getByTestId('row-0-cell-name')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-name')).toHaveTextContent('Table');
    expect(screen.getByTestId('row-0-cell-type')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-type')).toHaveTextContent('string');
    expect(screen.getByTestId('row-0-cell-required')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-required')).toHaveTextContent('true');
    expect(screen.getByTestId('row-0-cell-defaultValue')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-defaultValue')).toHaveTextContent('');
    expect(screen.getByTestId('row-0-cell-description')).toBeInTheDocument();
    expect(screen.getByTestId('row-0-cell-description')).toHaveTextContent('The name of the DynamoDB table.');
    expect(screen.getByTestId('row-1-cell-name')).toBeInTheDocument();
    expect(screen.getByTestId('row-1-cell-name')).toHaveTextContent('Key');
    expect(screen.getByTestId('row-1-cell-type')).toBeInTheDocument();
    expect(screen.getByTestId('row-1-cell-type')).toHaveTextContent('string');
    expect(screen.getByTestId('row-1-cell-required')).toBeInTheDocument();
    expect(screen.getByTestId('row-1-cell-required')).toHaveTextContent('false');
    expect(screen.getByTestId('row-1-cell-defaultValue')).toBeInTheDocument();
    expect(screen.getByTestId('row-1-cell-defaultValue')).toHaveTextContent('amazonaws.com');
    expect(screen.getByTestId('row-1-cell-description')).toBeInTheDocument();
    expect(screen.getByTestId('row-1-cell-description')).toHaveTextContent('New key');
  });
});

describe('Unknown tile', () => {
  const tile: ITile = {
    type: 'tile-type',
    name: 'tile-name',
    title: 'tile-title',
    description: 'tile-description',
    tags: ['tag1', 'tag2'],
    headerTags: ['header-tag1', 'header-tag2'],
    rawObject: {},
  };

  // it suppresses error in the console.log since it is expected
  const consoleErrorFn = jest.spyOn(console, 'error').mockImplementation(() => jest.fn());

  afterAll(() => {
    consoleErrorFn.mockRestore()
  });

  it('fires error for property modal', () => {
    
    expect(() => render(<PropertiesModal tile={tile} isModalOpen={true} onClose={jest.fn()} />)).toThrow(
      'Unknown CatalogKind during rendering modal: tile-type',
    );
  });
});
