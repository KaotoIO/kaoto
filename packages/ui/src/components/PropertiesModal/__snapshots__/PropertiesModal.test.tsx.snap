// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component tile renders property modal correctly 1`] = `
<body
  class="pf-v5-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-v5-c-backdrop"
    >
      <div
        class="pf-v5-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-2"
          aria-labelledby="pf-modal-part-1"
          aria-modal="true"
          class="pf-v5-c-modal-box properties-modal"
          data-ouia-component-id="BasicModal"
          data-ouia-component-type="PF5/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-0"
          role="dialog"
        >
          <div
            class="pf-v5-c-modal-box__close"
          >
            <button
              aria-disabled="false"
              aria-label="Close"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="BasicModal-ModalBoxCloseButton"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </div>
          <header
            class="pf-v5-c-modal-box__header"
          >
            <h1
              class="pf-v5-c-modal-box__title"
              id="pf-modal-part-1"
            >
              <span
                class="pf-v5-c-modal-box__title-text"
              >
                Atom
              </span>
            </h1>
          </header>
          <div
            class="pf-v5-c-modal-box__body"
            id="pf-modal-part-2"
          >
            <p
              data-testid="properties-modal-table-description"
            >
              Poll Atom RSS feeds.
            </p>
            <table
              aria-label="Simple table"
              class="pf-v5-c-table pf-m-grid-md pf-m-compact"
              data-ouia-component-id="OUIA-Generated-Table-1"
              data-ouia-component-type="PF5/Table"
              data-ouia-safe="true"
              role="grid"
            >
              <caption
                class="pf-v5-c-table__caption"
                data-testid="properties-modal-table-caption"
              >
                Available properties (1)
              </caption>
              <thead
                class="pf-v5-c-table__thead"
              >
                <tr
                  class="pf-v5-c-table__tr"
                  data-ouia-component-id="OUIA-Generated-TableRow-1"
                  data-ouia-component-type="PF5/TableRow"
                  data-ouia-safe="true"
                >
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-name"
                    scope="col"
                    tabindex="-1"
                  >
                    name
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-type"
                    scope="col"
                    tabindex="-1"
                  >
                    type
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-kind"
                    scope="col"
                    tabindex="-1"
                  >
                    kind
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-required"
                    scope="col"
                    tabindex="-1"
                  >
                    required
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-defaultValue"
                    scope="col"
                    tabindex="-1"
                  >
                    defaultValue
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-description"
                    scope="col"
                    tabindex="-1"
                  >
                    description
                  </th>
                </tr>
              </thead>
              <tbody
                class="pf-v5-c-table__tbody"
                role="rowgroup"
              >
                <tr
                  class="pf-v5-c-table__tr"
                  data-ouia-component-id="OUIA-Generated-TableRow-2"
                  data-ouia-component-type="PF5/TableRow"
                  data-ouia-safe="true"
                  data-testid="row-0"
                >
                  <td
                    class="pf-v5-c-table__td"
                    data-label="name"
                    data-testid="row-0-cell-name"
                    tabindex="-1"
                  >
                    Bridge Error Handler
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="type"
                    data-testid="row-0-cell-type"
                    tabindex="-1"
                  >
                    boolean
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="kind"
                    data-testid="row-0-cell-kind"
                    tabindex="-1"
                  >
                    property
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="required"
                    data-testid="row-0-cell-required"
                    tabindex="-1"
                  >
                    false
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="defaultValue"
                    data-testid="row-0-cell-defaultValue"
                    tabindex="-1"
                  >
                    false
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="description"
                    data-testid="row-0-cell-description"
                    tabindex="-1"
                  >
                    Allows for bridging the consumer to the Camel
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Kamelet tile renders property modal correctly 1`] = `
<body
  class="pf-v5-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-v5-c-backdrop"
    >
      <div
        class="pf-v5-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-4"
          aria-labelledby="pf-modal-part-3"
          aria-modal="true"
          class="pf-v5-c-modal-box properties-modal"
          data-ouia-component-id="BasicModal"
          data-ouia-component-type="PF5/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-2"
          role="dialog"
        >
          <div
            class="pf-v5-c-modal-box__close"
          >
            <button
              aria-disabled="false"
              aria-label="Close"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="BasicModal-ModalBoxCloseButton"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </div>
          <header
            class="pf-v5-c-modal-box__header"
          >
            <h1
              class="pf-v5-c-modal-box__title"
              id="pf-modal-part-3"
            >
              <span
                class="pf-v5-c-modal-box__title-text"
              >
                AWS DynamoDB Streams Source
              </span>
            </h1>
          </header>
          <div
            class="pf-v5-c-modal-box__body"
            id="pf-modal-part-4"
          >
            <p
              data-testid="properties-modal-table-description"
            >
              Receive events from Amazon DynamoDB Streams.
            </p>
            <table
              aria-label="Simple table"
              class="pf-v5-c-table pf-m-grid-md pf-m-compact"
              data-ouia-component-id="OUIA-Generated-Table-3"
              data-ouia-component-type="PF5/Table"
              data-ouia-safe="true"
              role="grid"
            >
              <caption
                class="pf-v5-c-table__caption"
                data-testid="properties-modal-table-caption"
              >
                Available properties (2)
              </caption>
              <thead
                class="pf-v5-c-table__thead"
              >
                <tr
                  class="pf-v5-c-table__tr"
                  data-ouia-component-id="OUIA-Generated-TableRow-4"
                  data-ouia-component-type="PF5/TableRow"
                  data-ouia-safe="true"
                >
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-name"
                    scope="col"
                    tabindex="-1"
                  >
                    name
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-type"
                    scope="col"
                    tabindex="-1"
                  >
                    type
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-required"
                    scope="col"
                    tabindex="-1"
                  >
                    required
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-defaultValue"
                    scope="col"
                    tabindex="-1"
                  >
                    defaultValue
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-description"
                    scope="col"
                    tabindex="-1"
                  >
                    description
                  </th>
                </tr>
              </thead>
              <tbody
                class="pf-v5-c-table__tbody"
                role="rowgroup"
              >
                <tr
                  class="pf-v5-c-table__tr"
                  data-ouia-component-id="OUIA-Generated-TableRow-5"
                  data-ouia-component-type="PF5/TableRow"
                  data-ouia-safe="true"
                  data-testid="row-0"
                >
                  <td
                    class="pf-v5-c-table__td"
                    data-label="name"
                    data-testid="row-0-cell-name"
                    tabindex="-1"
                  >
                    Table
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="type"
                    data-testid="row-0-cell-type"
                    tabindex="-1"
                  >
                    string
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="required"
                    data-testid="row-0-cell-required"
                    tabindex="-1"
                  >
                    true
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="defaultValue"
                    data-testid="row-0-cell-defaultValue"
                    tabindex="-1"
                  />
                  <td
                    class="pf-v5-c-table__td"
                    data-label="description"
                    data-testid="row-0-cell-description"
                    tabindex="-1"
                  >
                    The name of the DynamoDB table.
                  </td>
                </tr>
                <tr
                  class="pf-v5-c-table__tr"
                  data-ouia-component-id="OUIA-Generated-TableRow-6"
                  data-ouia-component-type="PF5/TableRow"
                  data-ouia-safe="true"
                  data-testid="row-1"
                >
                  <td
                    class="pf-v5-c-table__td"
                    data-label="name"
                    data-testid="row-1-cell-name"
                    tabindex="-1"
                  >
                    Key
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="type"
                    data-testid="row-1-cell-type"
                    tabindex="-1"
                  >
                    string
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="required"
                    data-testid="row-1-cell-required"
                    tabindex="-1"
                  >
                    false
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="defaultValue"
                    data-testid="row-1-cell-defaultValue"
                    tabindex="-1"
                  >
                    amazonaws.com
                  </td>
                  <td
                    class="pf-v5-c-table__td"
                    data-label="description"
                    data-testid="row-1-cell-description"
                    tabindex="-1"
                  >
                    New key
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Model tile with empty properties renders property modal correctly 1`] = `
<body
  class="pf-v5-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-v5-c-backdrop"
    >
      <div
        class="pf-v5-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-3"
          aria-labelledby="pf-modal-part-2"
          aria-modal="true"
          class="pf-v5-c-modal-box properties-modal"
          data-ouia-component-id="BasicModal"
          data-ouia-component-type="PF5/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-1"
          role="dialog"
        >
          <div
            class="pf-v5-c-modal-box__close"
          >
            <button
              aria-disabled="false"
              aria-label="Close"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="BasicModal-ModalBoxCloseButton"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </div>
          <header
            class="pf-v5-c-modal-box__header"
          >
            <h1
              class="pf-v5-c-modal-box__title"
              id="pf-modal-part-2"
            >
              <span
                class="pf-v5-c-modal-box__title-text"
              >
                Api Key
              </span>
            </h1>
          </header>
          <div
            class="pf-v5-c-modal-box__body"
            id="pf-modal-part-3"
          >
            <p
              data-testid="properties-modal-table-description"
            >
              Rest security basic auth definition
            </p>
            <table
              aria-label="Simple table"
              class="pf-v5-c-table pf-m-grid-md pf-m-compact"
              data-ouia-component-id="OUIA-Generated-Table-2"
              data-ouia-component-type="PF5/Table"
              data-ouia-safe="true"
              role="grid"
            >
              <caption
                class="pf-v5-c-table__caption"
                data-testid="properties-modal-table-caption"
              >
                Available properties (0)
              </caption>
              <thead
                class="pf-v5-c-table__thead"
              >
                <tr
                  class="pf-v5-c-table__tr"
                  data-ouia-component-id="OUIA-Generated-TableRow-3"
                  data-ouia-component-type="PF5/TableRow"
                  data-ouia-safe="true"
                >
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-name"
                    scope="col"
                    tabindex="-1"
                  >
                    name
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-type"
                    scope="col"
                    tabindex="-1"
                  >
                    type
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-kind"
                    scope="col"
                    tabindex="-1"
                  >
                    kind
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-required"
                    scope="col"
                    tabindex="-1"
                  >
                    required
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-defaultValue"
                    scope="col"
                    tabindex="-1"
                  >
                    defaultValue
                  </th>
                  <th
                    class="pf-v5-c-table__th"
                    data-testid="header-description"
                    scope="col"
                    tabindex="-1"
                  >
                    description
                  </th>
                </tr>
              </thead>
              <tbody
                class="pf-v5-c-table__tbody"
                role="rowgroup"
              >
                <tr>
                  <td
                    colspan="8"
                  >
                    <div
                      class="pf-v5-l-bullseye"
                    >
                      <div
                        class="pf-v5-c-empty-state pf-m-sm"
                      >
                        <div
                          class="pf-v5-c-empty-state__content"
                        >
                          <div
                            class="pf-v5-c-empty-state__header"
                            data-testid="empty-state"
                          >
                            <div
                              class="pf-v5-c-empty-state__icon"
                            >
                              <svg
                                aria-hidden="true"
                                class="pf-v5-svg"
                                fill="currentColor"
                                height="1em"
                                role="img"
                                viewBox="0 0 512 512"
                                width="1em"
                              >
                                <path
                                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                                />
                              </svg>
                            </div>
                            <div
                              class="pf-v5-c-empty-state__title"
                            >
                              <h2
                                class="pf-v5-c-empty-state__title-text"
                              >
                                No properties found for apiKey
                              </h2>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
